datacube/model/__init__.py

```diff
@@ -614,13 +614,16 @@ class MetadataType:
     def __init__(
         self,
         definition: Mapping[str, Any],
-        dataset_search_fields: Mapping[str, Field] | None = None,
+        search_field_extractor: Callable[[Mapping[str, Any]], Mapping[str, Field]]
```

Comment:
I like the `Mapping` in the return type, indicates that this stuff is not supposed to be changed. Should the return type of `get_dataset_fields` be changed to Mapping as well?